fiftyone/core/odm/database.py

```diff
@@ -7,42 +7,130 @@
+    URI. Otherwise, a :class:`fiftyone.core.service.DatabaseService` is
+    created.
+
+    Args:
+        config: a :class:`fiftyone.core.config.FiftyOneConfig`
+
+    Raises:
+        ConnectionError: if a connection to ``mongod`` could not be established
+        FiftyOneConfigError: if ``fiftyone.config.database_uri`` is not
+            defined and ``mongod`` could not be found
+        ServiceExecutableNotFound: if
+            :class:`fiftyone.core.service.DatabaseService` startup was
+            attempted, but ``mongod`` was not found in :mod:`fiftyone.db.bin`
+        RuntimeError: if the ``mongod`` found does not meet FiftyOne's
+            requirements, or validation could not occur
+    """
+    global _connection_kwargs
+
+    if config.database_uri is not None:
+        _connection_kwargs["host"] = config.database_uri
+    else:
+        global _db_service
+
+        if _db_service is not None:
+            return
+
+        if os.environ.get("FIFTYONE_DISABLE_SERVICES", False):
+            return
+
+        try:
+            _db_service = fos.DatabaseService()
+            _connection_kwargs["port"] = _db_service.port
+
+        except fos.ServiceExecutableNotFound as error:
+            if not fou.is_arm_mac():
+                raise error
+
+            raise FiftyOneConfigError(
+                "MongoDB is not yet supported on Apple Silicon Macs. Please"
+                "define a `database_uri` in your"
+                "`fiftyone.core.config.FiftyOneConfig` to define a connection"
+                "to your own MongoDB instance or cluster"
+            )
+
+    global _client
+
+    _client = pymongo.MongoClient(**_connection_kwargs)
+    _validate_db_version(config, _client)
+
+    connect(foc.DEFAULT_DATABASE, **_connection_kwargs)
+
+
 def _connect():
     global _client
     if _client is None:
-        connect(foc.DEFAULT_DATABASE, port=_default_port)
-        _client = pymongo.MongoClient(port=_default_port)
+        global _connection_kwargs
+        _client = pymongo.MongoClient(**_connection_kwargs)
+        connect(foc.DEFAULT_DATABASE, **_connection_kwargs)
 
 
 def _async_connect():
     global _async_client
     if _async_client is None:
-        _async_client = motor.motor_tornado.MotorClient(
-            "localhost", _default_port
+        global _connection_kwargs
+        _async_client = motor.motor_tornado.MotorClient(**_connection_kwargs)
+
+
+def _validate_db_version(config, client):
+    try:
+        version = Version(client.server_info()["version"])
+    except Exception as e:
+        if isinstance(e, ServerSelectionTimeoutError):
+            raise ConnectionError("Could not connect to `mongod`") from e
+
+        raise RuntimeError("Failed to validate `mongod` version") from e
+
+    min_ver, max_ver = foc.MONGODB_VERSION_RANGE
+
+    if min_ver <= version < max_ver:
+        return
+
+    if version >= max_ver and config.database_uri is not None:
```

Comment:
Why is the "try upgrading FiftyOne" recommendation added specifically in the case when `database_uri is not None` here?

In general, is the idea that the user might have previously installed a newer version of FiftyOne that did an admin migration but they since reverted?

Reply:
Yes, if they connect to a mongo 5.0 instance (after we add support/transition) and they downgrade to 0.12.0 for some reason.

It's a small case, I suppose. In the near term, I think it will be more likely that their mongo version is too old.

Reply:
I see. Wouldn't this apply equally to FiftyOne-managed dbs and `database_uri` dbs though? Both should have up/down admin migrations run whenever the relevant fiftyone versions are installed.

I'm latching onto this primarily to make sure I understand the change, and this special case just confused me. If the "try upgrading" message were printed in all cases, I wouldn't have commented in the first place.

Reply:
This validation is using the version of the mongo binary. `fiftyone-db` should always be carrying the correct one.

This check has nothing to do with admin migrations. Upgrading and downgrading `fiftyone` should always upgrade/downgrade `fiftyone-db` accordingly, no?

Reply:
I see, I was thinking about things like `db.admin.command({"setFeatureCompatibilityVersion": "4.4"})` in our (only) admin migration, but that doesn't actually change the binary version, of course.